### PR TITLE
(ORC-503) Pass user_id, account_id, and rate-limiting values on

### PR DESCRIPTION
**Pull request by nickmarrone**
Other plugins need these values, and it is better to only parse the JWT once. 

Values are read from parsed JWT and set to the header once in SM-JWT, then can be used by other plugins later

Tested locally using httpbin.org/get

@simplymeasured/hooch

### CODE REVIEW

File: kong/constants.lua [hunk 1]
```diff
@@ -45,7 +45,11 @@ return {
     CONSUMER_USERNAME = "X-Consumer-Username",
     CREDENTIAL_USERNAME = "X-Credential-Username",
     RATELIMIT_LIMIT = "X-RateLimit-Limit",
-    RATELIMIT_REMAINING = "X-RateLimit-Remaining"
```
Comment (dempe) [inline kong/constants.lua]: We don't currently have a good method to deploy modified core Kong code.  Additionally, it would make updating to a new Kong version more difficult.

FYI:  [here's] (https://github.com/simplymeasured/chef/blob/master/cookbooks/sm-kong/recipes/custom_plugins.rb) how we're currently deploying Kong plugins to Kong nodes -- just copying them over to the "plugins" directory on each node.

Comment (dempe) [inline kong/constants.lua]: However, since this just contains nothing but constants, I'm wondering if we can't just deploy this like we do the plugins...

File: kong/plugins/sm-jwt/handler.lua [hunk 2]
```diff
@@ -10,7 +10,7 @@ local ngx_re_gmatch = ngx.re.gmatch
```
Comment (dempe) [inline kong/plugins/sm-jwt/handler.lua]: Just making a note:
- We'll need to start using `sm-jwt` over the regular `jwt` plugin.
  - Need to update chef to deploy these files.
  - Need to update kong-config
- This will need to updated to be compatible with the new Kong release.  We can just copy these functions over to the new 8.3 `jwt` plugin.

**Comment by dempe** [comment]
Looks good, but let's wait till I'm finished with the Kong version update.

File: kong/plugins/sm-jwt/handler.lua [hunk 3]
```diff
@@ -10,7 +10,7 @@ local ngx_re_gmatch = ngx.re.gmatch
 
 local JwtHandler = BasePlugin:extend()
 
-JwtHandler.PRIORITY = 1000
+JwtHandler.PRIORITY = 4000
```
Comment (ThorArakis) [inline kong/plugins/sm-jwt/handler.lua]: Why the change in priority?

**Comment by ThorArakis** [comment]
👍

File: kong/plugins/sm-jwt/handler.lua (hunk 3 above)
Comment (nickmarrone) [inline kong/plugins/sm-jwt/handler.lua]: Just wanted to make sure this went first even if other plugins are added.

**Comment by nickmarrone** [comment]
This has been updated by rebasing on kong. Tested locally and it works (sets user_id, account_id, month and minute ratelimiting).

File: kong/constants.lua [hunk 4]
```diff
@@ -28,7 +28,16 @@ return {
     RATELIMIT_REMAINING = "X-RateLimit-Remaining",
     CONSUMER_GROUPS = "X-Consumer-Groups",
     FORWARDED_HOST = "X-Forwarded-Host",
-    FORWARDED_PREFIX = "X-Forwarded-Prefix"
+    FORWARDED_PREFIX = "X-Forwarded-Prefix",
+    JWT_USER_ID = "X-Jwt-User-Id",
+    JWT_ACCOUNT_ID = "X-Jwt-Account-Id",
+    JWT_RATELIMIT_MONTH = "X-Jwt-Ratelimit-Month",
+    JWT_RATELIMIT_MINUTE = "X-Jwt-Ratelimit-Minute"
```
Comment (ThorArakis) [inline kong/constants.lua]: move these constants into the plugin for now.

Comment (nickmarrone) [inline kong/constants.lua]: The JWT plugin uses other constants as well, I will move them into the plugin.

Comment (dempe) [inline kong/constants.lua]: Nick, wait until you see our deployment process!!

File: kong/plugins/jwt/handler.lua [hunk 5]
```diff
@@ -128,6 +129,35 @@ function JwtHandler:access(conf)
   ngx.req.set_header(constants.HEADERS.CONSUMER_CUSTOM_ID, consumer.custom_id)
   ngx.req.set_header(constants.HEADERS.CONSUMER_USERNAME, consumer.username)
   ngx.ctx.authenticated_credential = jwt_secret
+
+  -- Set values used by other sm plugins
+
+  -- user_id
+  full_user_id = jwt.claims["user_id"]
```
Comment (dempe) [inline kong/plugins/jwt/handler.lua]: Any reason to not method-abstract these?

For starters...
```lua
local function is_blank(str)
    return str == nil or str == ''
end

local function parse_claim(jwt, key)
    local value = jwt.claims[key]

    if is_empty(value) then
        return responses.send_HTTP_BAD_REQUEST("Could not find "..key.." in JWT")
    end

    return value
end
```

File: kong/constants.lua (hunk 4 above)
Comment (dempe) [inline kong/constants.lua]: However, this might get ugly once we begin consuming these headers in other plugins.
If we can figure out lua module deployment, we might consider having an SM-constants module.

File: kong/plugins/jwt/handler.lua [hunk 6]
```diff
@@ -128,6 +129,35 @@ function JwtHandler:access(conf)
...
+
+  -- account_id
+  account_id = jwt.claims["account_id"]
+  if not (account_id == nil or account_id == "") then
+    ngx.req.set_header(constants.HEADERS.JWT_ACCOUNT_ID, account_id)
+  end
+
+  -- rate_limit
+  ratelimit = jwt.claims["ratelimit"]
+  if not (ratelimit == nil or ratelimit == "") then
```
Comment (dempe) [inline kong/plugins/jwt/handler.lua]: Minor, but to avoid unnecessary indentation, I'm a fan of returning early.  
I.e., 
```lua 
if rate_limit == nil then 
    return 
end
```
NOTE:  `ratelimit == ""`, shouldn't be necessary as it's not a string.

Comment (nickmarrone) [inline kong/plugins/jwt/handler.lua]: If there is an error then I agree, returning early is a clean way of handling it. However, it is not an error to be missing the rate-limiting values. While the code would work right now, later on someone might add code below the rate-limit processing code which would be missed if rate-limiting is not present.

File: kong/plugins/jwt/handler.lua (hunk 5 above)
Comment (nickmarrone) [inline kong/plugins/jwt/handler.lua]: Good plan. is_blank works well for my purposes.

**Comment by ThorArakis** [comment]
👍

File: kong/plugins/jwt/handler.lua [hunk 7]
```diff
@@ -124,10 +136,35 @@ function JwtHandler:access(conf)
...
+
+  -- account_id
+  if not is_blank(jwt.claims["account_id"]) then
+    ngx.req.set_header(JWT_ACCOUNT_ID, jwt.claims["account_id"])
+  end
+
+  -- rate_limit
+  ratelimit = jwt.claims["ratelimit"]
+  if not (ratelimit == nil) then
+      if not (ratelimit["month"] == nil) then
```
Comment (dempe) [inline kong/plugins/jwt/handler.lua]: These are strings, so you might wanna use the `is_blank` function!

File: kong/plugins/jwt/handler.lua (hunk 6 above)
Comment (dempe) [inline kong/plugins/jwt/handler.lua]: Ideally, the `rate_limit` header setting would be in its own method, so you wouldn't have to worry about that. :P

File: kong/plugins/jwt/handler.lua (hunk 7 above)
Comment (nickmarrone) [inline kong/plugins/jwt/handler.lua]: They are actually ints!

**Comment by dempe** [comment]
👍
